pymbolic/typing.py

```diff
@@ -1,14 +1,46 @@
-from pymbolic.primitives import Expression
-from numbers import Number
-from typing import Union
+from __future__ import annotations
 
-try:
+
+from typing import TYPE_CHECKING, TypeVar, Union, Tuple
+from typing_extensions import TypeAlias
+
+if TYPE_CHECKING:
+    from pymbolic.primitives import Expression
+
+
+if TYPE_CHECKING:
+    # Yes, type-checking pymbolic will require numpy. That's OK.
     import numpy as np
-except ImportError:
-    BoolT = bool
-else:
     BoolT = Union[bool, np.bool_]
+    ScalarT: TypeAlias = Union[int, BoolT, float, np.generic]
+else:
+    try:
+        import numpy as np
+    except ImportError:
+        BoolT = bool
+        ScalarT: TypeAlias = Union[int, BoolT, float]
```

Comment:
Not sure it's worth the trouble, but.. complex?

Reply:
Good point, reworked quite a bit.

Reply:
(Leaving this unresolved, so that you can take a look should you so choose.)

Reply:
It looks nice enough now, no further comments! :D

Reply:
Based on experience with loopy, I've reworked these types yet again, I hope they're still to your liking. :slightly_smiling_face:

Reply:
Yeah, that looks fine to me! I mostly mentioned `complex` because of Helmholtz (and possibly other things in `sumpy` that don't mind being complex).